Add support for configurable paginator on top of search results.

EreMaijala opened the pull request:
Extracted from #2929.

EreMaijala committed:
Fix email-based catalog login from record holdings tab. (#3030)

EreMaijala committed:
Add support for configurable paginator on top of search results.

EreMaijala committed:
Add pagination-top class and tests. Cleanup.

EreMaijala commented:
Mink tests are passing for me.

demiankatz commented on config/vufind/searches.ini:
```diff
@@ -153,6 +153,13 @@ retain_filters_by_default = true
+; empty string or false  No top pagination (default)
+; simple                 Simple next/prev button pagination
+; full                   Full pagination alike to the one at the bottom of results
+;top_paginator = simple
```
Do we also need to add this setting to all the other search-related .ini files? Has it been tested to work in the context of other backends?

Is it worth considering a bottom_paginator setting, so both paginators can be configured independently? Maybe that adds too much complexity (since we would probably want to validate that the user doesn't turn them both off and render the result set impossible to navigate), but it seems conceivable that somebody might want to have ONLY a top paginator, and there's currently no way to configure that.

EreMaijala replied:
I've now tested with Blender, EDS and Primo, and can't see any reason why it wouldn't work with others as well, so I went ahead and added it, and tried to follow the existing convention in each files regarding comments.

If you really want to, you can at least hide the bottom navigation with css. But I wouldn't want to encourage that, since I can't see that being a good idea in normal cases.

demiankatz commented on themes/bootstrap3/templates/search/controls/showing.phtml:
```diff
@@ -10,21 +10,25 @@
+  // Use simple format if search-heading is set, search type is not basic or we have a simple top paginator enabled:
+  $topPaginator = $this->results ? $this->results->getOptions()->getTopPaginatorStyle() : '';
+  if ($this->slot('search-heading')->isset() || $this->params->getSearchType() !== 'basic' || 'simple' === $topPaginator) {
+    $showingResults = !empty($this->skipTotalCount)
```
Isn't `$showingResults` _already_ set to this value at this point in the code due to lines 9-12 above? Since this if..else guarantees that the variable will be set to something, can we just remove those earlier lines?

EreMaijala replied:
Oops, yes, though the condition wasn't quite the same! Now cleaned up.

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/BasicSearchTest.php:
```diff
@@ -58,4 +58,97 @@ public function testOutOfBoundsPage()
+        $secondPage->click();
+        $this->waitForPageLoad($page);
+        // Prev page now present:
+        $this->findCss($page, '.search-header .pagination-simple .page-prev');
```
Is it worth clicking on the prev button and asserting that it returns you to the previous page, so that both buttons are tested?

EreMaijala replied:
Good idea, now added.

demiankatz reviewed (changes requested):
Thanks, @EreMaijala -- a few comments/questions.

EreMaijala committed:
Introduce results-list template between results and list templates. (#3032)

This paves way for extended use (i.e. the possibility of rendering search results without the surrounding controls).

EreMaijala committed:
Add new setting to more ini files.

EreMaijala committed:
Remove useless code.

EreMaijala committed:
Improve tests.

demiankatz committed:
Merge branch 'dev-10.0' into dev-10.0-top-navigation

demiankatz commented on module/VuFind/tests/integration-tests/src/VuFindTest/Mink/BasicSearchTest.php:
```diff
@@ -58,4 +58,126 @@ public function testOutOfBoundsPage()
+        $page = $session->getPage();
+
+        // No paginator unless configured:
+        $this->unFindCss($page, '.search-header .pagination');
```
Why are you checking for `.search-header .pagination` here but using `.pagination-top` when you actually interact with the page? It seems like it might be worth making the selectors more consistent (or at least adding additional find/unfind checks) to make it more clear that the test is proving what it says that it's proving.

EreMaijala replied:
Ah, that check was added before I added the pagination-top class. Now simplified.

demiankatz reviewed (changes requested):
Thanks for the progress, and I agree that adding more configuration to hide the bottom paginator is probably not necessary at this point (and could be added in future if there's a compelling need). Just one small question that arose when I looked more closely at the Mink tests.

EreMaijala committed:
Simplify the check.

demiankatz reviewed (approved):
Thanks, @EreMaijala -- looks good to me now, and all Mink tests are still passing in my environment.
